daemon/src/model/cfd.rs

```diff
@@ -1276,6 +1279,273 @@ impl Cfd {
     }
 }
 
+trait Aggregate {
+    type Item;
+
+    fn version(&self) -> u64;
+    fn apply(self, evt: &Self::Item) -> Self
+    where
+        Self: Sized;
+}
+
+pub mod taker {
+
+    use super::*;
+
+    /// Models the cfd state
+    ///
+    /// Upon `Command`s, that are reaction to something happening in the system, we decide to
+    /// produce `Event`s that are saved in the database. After saving an `Event` in the database
+    /// we apply the event to the aggregate producing a new aggregate (representing the latest state
+    /// `version`). To bring a cfd into a certain state version we load all events from the
+    /// database and apply them in order (order by version).
+    pub struct CfdAggregate {
+        version: u64,
+
+        // static
+        uuid: Uuid,
+        position: Position,
+        initial_price: Price,
+        leverage: Leverage,
+        settlement_time_interval_hours: Duration,
+        quantity: Usd,
+
+        // TODO: plug in once we model this
+        // pub counterparty_network_identity: Uuid,
+
+        // dynamic (based on events)
+        dlc: Option<Dlc>,
+        cet: Option<Cet>,
+        attestation: Option<Attestation>,
+
+        revoke_commit: Option<RevokedCommit>,
+
+        // TODO: Proabbly better to bundle this in some structure?
+        lock_finality: bool,
+        commit_finality: bool,
+        refund_finality: bool,
+        cet_finality: bool,
+        collaborative_settlement_finality: bool,
+
+        cet_timelock_expired: bool,
+        refund_timelock_expired: bool,
+
+        // TODO: Can we map to cfd state for now to have contained changes in the legacy code?
+        cfd_state: CfdState,
+    }
+
+    impl CfdAggregate {
+        pub fn new(
+            uuid: Uuid,
+            position: Position,
+            initial_price: Price,
+            leverage: Leverage,
+            settlement_time_interval_hours: Duration,
+            quantity: Usd,
+        ) -> Self {
+            CfdAggregate {
+                version: 1,
+                uuid,
+                position,
+                initial_price,
+                leverage,
+                settlement_time_interval_hours,
+                quantity,
+                dlc: None,
+                cet: None,
+                attestation: None,
+                revoke_commit: None,
+                lock_finality: false,
+                commit_finality: false,
+                refund_finality: false,
+                cet_finality: false,
+                collaborative_settlement_finality: false,
+                cet_timelock_expired: false,
+                refund_timelock_expired: false,
+                cfd_state: CfdState::outgoing_order_request(),
+            }
+        }
+
+        pub fn start_contract_setup(&self) {
+            todo!(
+                "should this advance state? if it does not then there is not point in having this"
```

Comment:
It shouldn't advance state but the point of this is to push this logic into the domain model.

Additionally, starting a contract setup is not valid in all states hence there is value in having this on here, esp. once the state machine for performing the contract setup.

Lastly, having this on here allows us to access private data nicely that will be needed for contract setup instead of exposing dumb getters and composing things outside of the domain model.